kv/kvserver: TestLeasePreferencesDuringOutage failed

kv/kvserver.TestLeasePreferencesDuringOutage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10107163?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10107163?buildTab=artifacts#/) on release-22.2 @ [ca300add723ec3cc4404cacd56ee85e74d0b29ed](https://github.com/cockroachdb/cockroach/commits/ca300add723ec3cc4404cacd56ee85e74d0b29ed):


```
=== RUN   TestLeasePreferencesDuringOutage
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeasePreferencesDuringOutage2126522253
    test_log_scope.go:79: use -show-logs to present logs inline
*
* INFO: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
*
    client_lease_test.go:1062: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2880/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/client_lease_test.go:1062
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 5
        	Test:       	TestLeasePreferencesDuringOutage
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeasePreferencesDuringOutage2126522253
--- FAIL: TestLeasePreferencesDuringOutage (4.21s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeasePreferencesDuringOutage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27979

--- a/pkg/kv/kvserver/client_lease_test.go
+++ b/pkg/kv/kvserver/client_lease_test.go
@@ -16,6 +16,7 @@ import (
 	"math"
 	"runtime"
 	"strconv"
+	"strings"
 	"sync"
 	"sync/atomic"
 	"testing"
@@ -991,8 +992,8 @@ func TestLeasePreferencesDuringOutage(t *testing.T) {
 	tc.StopServer(1)
 	tc.StopServer(2)
 
-	wait := func(duration int64) {
-		manualClock.Increment(duration)
+	wait := func(duration time.Duration) {
+		manualClock.Increment(duration.Nanoseconds())
 		// Gossip and heartbeat all the live stores, we do this manually otherwise the
 		// allocator on server 0 may see everyone as temporarily dead due to the
 		// clock move above.
@@ -1003,7 +1004,7 @@ func TestLeasePreferencesDuringOutage(t *testing.T) {
 	}
 	// We need to wait until 2 and 3 are considered to be dead.
 	timeUntilStoreDead := storepool.TimeUntilStoreDead.Get(&tc.GetFirstStoreFromServer(t, 0).GetStoreConfig().Settings.SV)
-	wait(timeUntilStoreDead.Nanoseconds())
+	wait(timeUntilStoreDead)
 
 	checkDead := func(store *kvserver.Store, storeIdx int) error {
 		if dead, timetoDie, err := store.GetStoreConfig().StorePool.IsDead(
@@ -1011,11 +1012,11 @@ func TestLeasePreferencesDuringOutage(t *testing.T) {
 			// Sometimes a gossip update arrives right after server shutdown and
 			// after we manually moved the time, so move it again.
 			if err == nil {
-				wait(timetoDie.Nanoseconds())
+				wait(timetoDie)
 			}
 			// NB: errors.Wrapf(nil, ...) returns nil.
 			// nolint:errwrap
-			return errors.Errorf("expected server 2 to be dead, instead err=%v, dead=%v", err, dead)
+			return errors.Errorf("expected server %d to be dead, instead err=%v, dead=%v", storeIdx, err, dead)
 		}
 		return nil
 	}
@@ -1038,7 +1039,7 @@ func TestLeasePreferencesDuringOutage(t *testing.T) {
 	_, _, enqueueError := tc.GetFirstStoreFromServer(t, 0).
 		Enqueue(ctx, "replicate", repl, true /* skipShouldQueue */, false /* async */)
 
-	require.NoError(t, enqueueError)
+	require.NoError(t, enqueueError, "failed to enqueue replica for replication")
 
 	var newLeaseHolder roachpb.ReplicationTarget
 	testutils.SucceedsSoon(t, func() error {
@@ -1062,8 +1063,28 @@ func TestLeasePreferencesDuringOutage(t *testing.T) {
 		require.NotEqual(t, "sf", dc)
 	}
 	history := repl.GetLeaseHistory()
-	// make sure we see the eu node as a lease holder in the second to last position.
-	require.Equal(t, tc.Target(0).NodeID, history[len(history)-2].Replica.NodeID)
+	// Make sure we see the eu node as a lease holder in the second to last
+	// leaseholder change.
+	// Since we can have expiration and epoch based leases at the tail of the
+	// history, we need to ignore them together if they originate from the same
+	// leaseholder.
+	nextNodeID := history[len(history)-1].Replica.NodeID
+	lastMove := len(history) - 2
+	for ; lastMove >= 0; lastMove-- {
+		if history[lastMove].Replica.NodeID != nextNodeID {
+			break
+		}
+	}
+	lastMove++
+	var leasesMsg []string
+	for _, h := range history {
+		leasesMsg = append(leasesMsg, h.String())
+	}
+	leaseHistory := strings.Join(leasesMsg, ", ")
+	require.Greater(t, lastMove, 0,
+		"must have at least one leaseholder change in history (lease history: %s)", leaseHistory)
+	require.Equal(t, tc.Target(0).NodeID, history[lastMove-1].Replica.NodeID,
+		"node id prior to last lease move (lease history: %s)", leaseHistory)
 }
 
 // This test verifies that when a node starts flapping its liveness, all leases

